### PR TITLE
Add ARP API in order to let clients apply ARPs to a set of attributes

### PR DESCRIPTION
**Pull request by arothuis**
This implements OpenConext/OpenConext-Profile#51.

Relates to OpenConext/OpenConext-Profile#70.

### CODE REVIEW

File: src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+     * @param Request $request
+     * @return JsonResponse
+     */
+    public function applyArpAction(Request $request)
+    {
+        if (!$this->authorizationChecker->isGranted('ROLE_API_USER_PROFILE')) {
+            throw new ApiAccessDeniedHttpException('Access to the ARP API requires the role ROLE_API_USER_PROFILE');
+        }
+
+        $body = json_decode($request->getContent(), true);
```
Comment (arothuis) [inline src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php]: We have a JSON helper class for this, but it converts JSON to objects. As the ARP enforcer expects arrays, using a native json_decode with some validation afterwards removes weird mapping logic from objects to arrays.

**Comment by arothuis** [comment]
@rjkip, 🍗

File: src/OpenConext/EngineBlockBundle/Exception/InvalidArgumentException.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace OpenConext\EngineBlockBundle\Exception;
+
+use InvalidArgumentException as CoreInvalidArgumentException;
+
+class InvalidArgumentException extends CoreInvalidArgumentException
+{
+    public static function invalidType($expectedType, $propertyPath, $parameter)
```
Comment (rjkip) [inline src/OpenConext/EngineBlockBundle/Exception/InvalidArgumentException.php]: Why was this needed?

File: src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    /**
+     * @var EngineBlock_Arp_AttributeReleasePolicyEnforcer
+     */
+    private $arpEnforcer;
+
+    public function __construct(
+        AuthorizationCheckerInterface $authorizationChecker,
+        MetadataService $metadataService,
+        EngineBlock_Arp_AttributeReleasePolicyEnforcer $arpEnforcer
```
Comment (rjkip) [inline src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php]: Is it project style to not alias classes with underscore namespacing?

File: src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    public function applyArpAction(Request $request)
+    {
+        if (!$this->authorizationChecker->isGranted('ROLE_API_USER_PROFILE')) {
+            throw new ApiAccessDeniedHttpException('Access to the ARP API requires the role ROLE_API_USER_PROFILE');
+        }
+
+        $body = json_decode($request->getContent(), true);
+
+        if (!is_array($body)) {
+            throw new BadApiRequestHttpException('Unrecognized structure for JSON: ' . var_export($body, true));
```
Comment (rjkip) [inline src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php]: This log message doesn't really help me towards a solution; what is wrong with the JSON structure, what is expected? What about `Root value submitted in request must be an array, got <type>`? Also, I'm not that fond of logging entire request bodies (we are dealing with a trusted user here, but still) and, implicitly, newlines.

File: src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php (hunk 1 above)
Comment (rjkip) [inline src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php]: Hmm, why not create a helper that returns arrays?

File: src/OpenConext/EngineBlock/Service/MetadataService.php [hunk 5]
```diff
@@ -49,4 +50,13 @@ public function findServiceProvider(EntityId $entityId)
 
         return $serviceProvider;
     }
+
+    /**
+     * @param ServiceProvider $serviceProvider
+     * @return null|AttributeReleasePolicy
+     */
+    public function findArpFor(ServiceProvider $serviceProvider)
```
Comment (rjkip) [inline src/OpenConext/EngineBlock/Service/MetadataService.php]: Why not accept an ID? Now we need to retrieve the service provider itself before retrieving its ARP.

File: src/OpenConext/EngineBlockBundle/Resources/config/routing_api.yml [hunk 6]
```diff
@@ -25,3 +25,9 @@ api_metadata_idp:
     methods: [GET]
     defaults:
         _controller: engineblock.controller.api.metadata:idpAction
+
+api_attribute_release_policy:
+    path:   /arp
+    methods: [POST]
```
Comment (thijskh) [inline src/OpenConext/EngineBlockBundle/Resources/config/routing_api.yml]: Why is this a POST? Does it modify any state?

File: src/OpenConext/EngineBlockBundle/Exception/InvalidArgumentException.php (hunk 2 above)
Comment (arothuis) [inline src/OpenConext/EngineBlockBundle/Exception/InvalidArgumentException.php]: Because it wasn't there, yet there was a class depending on it. I got an error when the Exception could not be thrown.

File: src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php (hunk 3 above)
Comment (arothuis) [inline src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php]: Thanks for noticing this. Considering we have the old and the new EB, I will look into what to do regarding underscore namespacing for older components.

File: src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php (hunk 4 above)
Comment (arothuis) [inline src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php]: This was a mistake on my part. I used it when debugging / exploring the request handling and I did not create a proper Exception message afterwards.

File: src/OpenConext/EngineBlock/Service/MetadataService.php (hunk 5 above)
Comment (arothuis) [inline src/OpenConext/EngineBlock/Service/MetadataService.php]: Nice idea, that would reduce a call to the service. We could give it the entityId and let this method deal with the rest.

File: src/OpenConext/EngineBlockBundle/Resources/config/routing_api.yml (hunk 6 above)
Comment (arothuis) [inline src/OpenConext/EngineBlockBundle/Resources/config/routing_api.yml]: After giving it some more thoughts and exchanging thoughts with colleagues, I think POST is appropriate. In principle, GET is the correct thing to do when retrieving a resource. Here, we are not retrieving a resource but we are doing a non-idempotent RPC-call: clients are basically saying "EngineBlock, please process this data for us and return the result".

Comment (thijskh) [inline src/OpenConext/EngineBlockBundle/Resources/config/routing_api.yml]: :+1:

File: src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php (hunk 3 above)
Comment (arothuis) [inline src/OpenConext/EngineBlockBundle/Controller/Api/AttributeReleasePolicyController.php]: For future reference: we are still using underscore namespacing for using older components. I think it makes sense to make it as explicit as possible whether you're dealing with some kind of compatibility component or not.

**Comment by arothuis** [comment]
The meat of this PR, the ARP enforcer, is already unit tested. As it's an RPC-type API, it would make sense to create a component test at the controller-level. PHPUnit with some mocks would suffice IMHO.

File: src/OpenConext/EngineBlock/Service/MetadataService.php [hunk 7]
```diff
@@ -49,4 +50,20 @@ public function findServiceProvider(EntityId $entityId)
...
+    /**
+     * @param EntityId $entityId
+     * @return null|AttributeReleasePolicy
+     * @internal param ServiceProvider $serviceProvider
+     */
+    public function findArpForServiceProviderByEntityId(EntityId $entityId)
+    {
+        $serviceProvider = $this->findServiceProvider($entityId);
+
+        if ($serviceProvider === null) {
```
Comment (arothuis) [inline src/OpenConext/EngineBlock/Service/MetadataService.php]: This is an edge case: our client (Profile) sends us a list of entityIds he retrieved earlier, but in the meantime the service provider could have been removed.

The code here means that if we cannot find a service provider, we assume there is no ARP. This causes us to show all attributes being released for this service provider, even though that might not have been the case. How should EngineBlock deal with this?

Thoughts, @thijskh?

Comment (thijskh) [inline src/OpenConext/EngineBlock/Service/MetadataService.php]: So this happens when there's a race condition: the SP was present when the list of SP's was retrieved, but removed before the ARP was applied. This is very rare and may in fact never happen in practice. So I'm fine with a simple approach that assumes no ARP.

File: src/OpenConext/EngineBlock/Service/MetadataService.php [hunk 8]
```diff
@@ -49,4 +50,20 @@ public function findServiceProvider(EntityId $entityId)
 
         return $serviceProvider;
     }
+
+    /**
+     * @param EntityId $entityId
+     * @return null|AttributeReleasePolicy
+     * @internal param ServiceProvider $serviceProvider
```
Comment (arothuis) [inline src/OpenConext/EngineBlock/Service/MetadataService.php]: Remove.

File: tests/unit/OpenConext/EngineBlockBundle/Controller/AttributeReleasePolicyControllerTest.php [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+    {
+        $this->expectException(BadApiRequestHttpException::class);
+        $this->expectExceptionMessage('expected decoded root value to be an array');
+
+        $request = $this->createRequestWithContent('null');
+
+        $arpController = new AttributeReleasePolicyController(
+            $this->mockAuthorizationChecker(),
+            $this->createDummyMetadataService(),
+            new EngineBlock_Arp_AttributeReleasePolicyEnforcer
```
Comment (rjkip) [inline tests/unit/OpenConext/EngineBlockBundle/Controller/AttributeReleasePolicyControllerTest.php]: We're this :ok_hand: far away from a unit test. But on the other hand, I'm interested in whether this works on the application level. Maybe it's worth looking into how feasible a WebTestCase is (timebox).

**Comment by arothuis** [comment]
I'm afraid we have to save the WebTestCase for another time. @rjkip, is this mergeable? :medal_military:
